src/com/liferay/portal/ejb/UserManagerImpl.java

```diff
@@ -751,6 +757,37 @@ private int _authenticate(
 		return authResult;
 	}
 
-	private static final Log _log = LogFactory.getLog(UserManagerImpl.class);
+	/**
+	 * This is a security measure for DoS and similar attacks in which the login
+	 * process will be delayed based on the number of times a specific user
+	 * fails to authenticate. Users are allowed to fail once without being
+	 * penalized for it. Different security strategies can be configured in
+	 * order to counter-attack these threats.
+	 * <p>
+	 * By default, the approach consists of retrieving the number of failed
+	 * login attempts and raise it to the power of 2. The result will represent
+	 * the seconds that the login process will wait before sending the error.
+	 * This way, the more times hackers attempt to authenticate incorrectly, the
+	 * more time they will have to wait to try it again.
+	 * 
+	 * @param failedAttempts
+	 *            - The number of times that a user has tried to log in and
+	 *            failed.
+	 */
+	private void delayLoginRequest(int failedAttempts) {

```

Comment:
This logic would be useful to encapsulate in a DelayService, so can be reused on other part of the system